packages/shared/src/util/features.ts

```diff
@@ -468,3 +468,207 @@ export function validateAndFixCondition(
+      for (const rule of env.rules || []) {
+        if (rule.prerequisites?.length) {
+          const rulePrerequisiteIds = rule.prerequisites.map((p) => p.id);
+          prerequisiteIds.push(...rulePrerequisiteIds);
+        }
+      }
+    }
+
+    for (const prerequisiteId of prerequisiteIds) {
+      const parentFeature = features.find((f) => f.id === prerequisiteId);
+      if (parentFeature && visit(parentFeature)[0])
+        return [true, prerequisiteId];
+    }
+
+    stack.delete(feature.id);
+    return [false, null];
+  };
+
+  return visit(newFeature);
+}
+
+type PrerequisiteState = "deterministic" | "conditional" | "cyclic";
+type PrerequisiteValue = string | null | undefined;
+export type PrerequisiteStateResult = {
+  state: PrerequisiteState;
+  value: PrerequisiteValue;
+};
+export function evaluatePrerequisiteState(
+  feature: FeatureInterface,
+  features: FeatureInterface[],
+  env: string,
+  skipRootConditions: boolean = false,
+  skipCyclicCheck: boolean = false
+): PrerequisiteStateResult {
+  let isTopLevel = true;
+  if (!skipCyclicCheck) {
+    if (isFeatureCyclic(feature, features)[0])
+      return { state: "cyclic", value: null };
+  }
+
+  const visit = (feature: FeatureInterface): PrerequisiteStateResult => {
+    // 1. Current environment toggles take priority
+    if (!feature.environmentSettings[env]) {
+      return { state: "deterministic", value: null };
+    }
+    if (!feature.environmentSettings[env].enabled) {
+      return { state: "deterministic", value: null };
+    }
+
+    // 2. Determine default feature state
+    //   - start with "deterministic" / defaultValue
+    //   - force "conditional" if there are rules
+    let state: PrerequisiteState = "deterministic";
+    let value: PrerequisiteValue = feature.defaultValue;
+    if (!skipRootConditions || !isTopLevel) {
+      if (
+        feature.environmentSettings[env].rules?.filter((r) => !!r.enabled)
+          ?.length
+      ) {
+        state = "conditional";
+        value = undefined;
+      }
+    }
+
+    // 3. If the feature has prerequisites, traverse all nodes (may override default state)
+    //  - if any are "off", the feature is "off"
+    //  - if any are "conditional", the feature is "conditional"
+    isTopLevel = false;
+    const prerequisites = feature.prerequisites || [];
+    for (const prerequisite of prerequisites) {
+      const prerequisiteFeature = features.find(
+        (f) => f.id === prerequisite.id
+      );
+      if (!prerequisiteFeature) {
+        // todo: consider returning info about missing feature
+        state = "deterministic";
+        value = null;
+        break;
+      }
+      const { state: prerequisiteState, value: prerequisiteValue } = visit(
+        prerequisiteFeature
+      );
+      if (prerequisiteState === "deterministic" && prerequisiteValue === null) {
+        // "off" by null. Any "off" prereq state overrides feature's default state (#2)
```

Comment:
Is this true in the general case?  What if you have a prerequisite on a rule with the condition: `{ value: {$exists: false} }`.  Then, the prereq being deterministic null should NOT override since the condition passes.

Reply:
this is only true for the current implementation of top-level prereqs (boolean, is true)

Reply:
replaced with evalCondition()